kroxylicious/src/test/java/io/kroxylicious/proxy/internal/SaslDecodePredicateTest.java

```diff
@@ -0,0 +1,130 @@
+/*
+ * Copyright Kroxylicious Authors.
+ *
+ * Licensed under the Apache Software License version 2.0, available at http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+package io.kroxylicious.proxy.internal;
+
+import org.apache.kafka.common.protocol.ApiKeys;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.EnumSource;
+
+import io.kroxylicious.proxy.internal.codec.DecodePredicate;
+
+import static org.junit.jupiter.api.Assertions.assertFalse;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+
+class SaslDecodePredicateTest {
+
+    private static final DecodePredicate TARGET_NOTHING = new DecodePredicate() {
+        @Override
+        public boolean shouldDecodeRequest(ApiKeys apiKey, short apiVersion) {
+            return false;
+        }
+
+        @Override
+        public boolean shouldDecodeResponse(ApiKeys apiKey, short apiVersion) {
+            return false;
+        }
+    };
+
+    private static final DecodePredicate TARGET_ALL = new DecodePredicate() {
+        @Override
+        public boolean shouldDecodeRequest(ApiKeys apiKey, short apiVersion) {
+            return true;
+        }
+
+        @Override
+        public boolean shouldDecodeResponse(ApiKeys apiKey, short apiVersion) {
+            return true;
+        }
+    };
+
+    private SaslDecodePredicate predicate;
+
+    @Test
+    public void testApiVersionAlwaysDecoded_SaslNotHandledAndBeforeDelegateSet() {
+        givenPredicateWithHandleSasl(false);
+        assertPredicateTargetsRequestKey(ApiKeys.API_VERSIONS);
+    }
+
+    @Test
+    public void testApiVersionAlwaysDecoded_SaslHandledAndBeforeDelegateSet() {
+        givenPredicateWithHandleSasl(true);
+        assertPredicateTargetsRequestKey(ApiKeys.API_VERSIONS);
+    }
+
+    @Test
+    public void testApiVersionAlwaysDecoded_SaslNotHandledAndDelegateSet() {
+        givenPredicateWithHandleSasl(false);
+        givenPredicateDelegateSet(TARGET_NOTHING);
```

Comment:
Similarly I'd probably go for a pair of 
- `givenPredicateTargetingNothing()` & `givenPredicateTargetingEverything()` 
- `givenDelegateTargetingNothing()` & `givenDelegateTargetingEverything()` 
- at a push `givenDelegate(TARGETING_NOTHING)` & `givenDelegate(TARGETING_EVERYTHING)`